### PR TITLE
Optimize withAll for primitive bag factories

### LINKED ISSUE #1372: Optimize withAll on MutableIntBagFactoryImpl and ImmutableIntBagFactoryImpl
The code currently in `withAll` on both mutable and immutable implementations calls `toArray` on `IntStream` which if the Stream is very large can be extremely expensive and won't take advantage of a Bag's ability to store duplicates as simply increments in a count.

Current mutable factory code:
```
public MutableIntBag withAll(IntStream items)
{
    return this.with(items.toArray());
}
```

Current immutable factory code:
```
public ImmutableIntBag withAll(IntStream items)
{
    return this.with(items.toArray());
}

```

It would be much better to simply iterate over the Stream and add the results to a HashBag, converting to an ImmutableBag in the immutable factory.

### PR DESCRIPTION
**Pull request by TheJavaGuy**
Please review this PR commit by commit starting from the earliest one. I’ve structured commits to be small and self-contained changes.

Every commit has a detailed message that explains why it was done that way. Of course, if you have additional questions I’m more than happy to answer them.

Closes #1372

### CODE REVIEW

**Comment by Desislav-Petrov** [comment]
hi @TheJavaGuy  - we usually squash all commits into a single one as per the guide here https://github.com/eclipse/eclipse-collections/blob/master/CONTRIBUTING.md

File: RELEASE_NOTE_DRAFT.md [hunk 1]
```diff
@@ -44,6 +44,7 @@ The Eclipse Collections team gives a huge thank you to everyone who participated
 * Optimized asParallel() for immutable sets in O(1) instead of O(n).
 * Optimized some implementations of toString() that were delegating to iterators.
 * Improved equals() performance for Set implementations.
+* Allowed creating immutable primitive bags with billions of elements.
```
Comment (donraab) [inline RELEASE_NOTE_DRAFT.md:47]: You can roll this change back as we accumulate all of the commits into the release notes draft before release.

File: unit-tests/src/test/java/org/eclipse/collections/impl/stream/PrimitiveStreamsTest.java [hunk 2]
```diff
@@ -51,6 +51,7 @@
 import org.eclipse.collections.impl.factory.primitive.LongSets;
 import org.eclipse.collections.impl.factory.primitive.LongStacks;
 import org.eclipse.collections.impl.list.primitive.IntInterval;
+import org.eclipse.collections.impl.list.primitive.LongInterval;
```
Comment (donraab) [inline unit-tests/src/test/java/org/eclipse/collections/impl/stream/PrimitiveStreamsTest.java:54]: Please update the copyright to "Copyright (c) 2022 The Bank of New York Mellon and others."

**Review by donraab — CHANGES_REQUESTED** [review]
Hi @TheJavaGuy, thank you for the contribution! I have a few changes requested inline. I appreciate all of the detail you have in the commits. It was helpful understanding the issues you encountered using toImmutable as I suggested in the issue. After you make the requested changes, you can squash the commits/rebase and I will review again and approve/merge assuming everything looks good.

File: unit-tests/src/test/java/org/eclipse/collections/impl/stream/PrimitiveStreamsTest.java (hunk 2 above)
Comment (TheJavaGuy) [reply unit-tests/src/test/java/org/eclipse/collections/impl/stream/PrimitiveStreamsTest.java:54]: Done

File: RELEASE_NOTE_DRAFT.md (hunk 1 above)
Comment (TheJavaGuy) [reply RELEASE_NOTE_DRAFT.md]: Done

File: eclipse-collections-code-generator/src/main/resources/impl/bag/immutable/immutablePrimitiveHashBag.stg [hunk 3]
```diff
@@ -79,6 +85,11 @@ final class Immutable<name>HashBag implements Immutable<name>Bag, Serializable
         return new Immutable<name>HashBag(elements);
     }
 
+    public static Immutable<name>HashBag newBagWith(Mutable<name>Bag elements)
+    {
+        return new Immutable<name>HashBag(elements);
```
[the remark this answers is not in the document]
Comment (TheJavaGuy) [reply eclipse-collections-code-generator/src/main/resources/impl/bag/immutable/immutablePrimitiveHashBag.stg:90]: Constructor already makes a defensive copy :) I’ve also added [a test](https://github.com/eclipse/eclipse-collections/pull/1376/files#diff-c9093a02cbeb8f81ad3d0de6405927d8929740da59d0d25bad9513d7f5dde52f) that demonstrates this.

**Comment by donraab** [comment]
Hi @TheJavaGuy, the changes look good. Please update the commit message to added closes with the issue number. Thanks!

**Review by donraab — APPROVED** [review]
LGTM! Thanks for the contribution @TheJavaGuy !

**Pull request merged**